### PR TITLE
[#631][PORT] [Slack Adapters] Add Slack Functional Test

### PR DESCRIPTION
**Pull request by iscaltritti**
Fixes # 631

## Description
This PR ports the Slack adapter functional test from PR# 3108 from botbuilder-dotnet.

## Specific Changes

  - Adds slacktestbot (migrated from [BotBuilder-Samples/samples/python/60.slack-adapter](https://github.com/microsoft/BotBuilder-Samples/tree/main/samples/python/60.slack-adapter))
  - Adds `test_slack_client.py` 
  - Adds `requirements.txt` in the _functional-tests_ library
  - Adds pipeline definition YAML to deploy the bot and run the test.

Note: SlackHistoryRetrieve port was not required, in DotNet it is used by JsonDeserealize to get the response message, but in Python, the implementation differs.

## Testing
The following images showcase the pipeline running the test successfully and how it looks on the bot through Slack:
image
image

### CODE REVIEW

File: libraries/functional-tests/tests/test_slack_client.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import datetime
+import time
+import aiounittest
+import requests
+
+
+class SlackClient(aiounittest.AsyncTestCase):
+    async def test_send_and_receive_slack_message(self):
+        # Arrange
+        # self._get_environment_vars()
```
Comment (sw-joelmut) [inline libraries/functional-tests/tests/test_slack_client.py]: This commented line should be removed

**Review by sw-joelmut — CHANGES_REQUESTED** [review]

**Review by sw-joelmut — APPROVED** [review]

File: pipelines/botbuilder-python-ci-slack-test.yml [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+    - pipelines/botbuilder-python-ci-slack-test.yml
+
+variables:
+  AppId: $(SlackTestBotAppId)
+  AppSecret: $(SlackTestBotAppSecret)
+  BotGroup: $(SlackTestBotBotGroup)
+  BotName: $(SlackTestBotBotName)
+  SlackBotToken: $(SlackTestBotSlackBotToken)
+  SlackClientSigningSecret: $(SlackTestBotSlackClientSigningSecret)
+  SlackVerificationToken: $(SlackTestBotSlackVerificationToken)
```
Comment (santgr11) [inline pipelines/botbuilder-python-ci-slack-test.yml:41]: Do we need these 3 here? I think that these are being used like that only for the script in line 87 and these variables are declared there

Comment (iscaltritti) [reply pipelines/botbuilder-python-ci-slack-test.yml:41]: They are used in the resource creation step, ported from the DotNet implementation.
They are indeed redundant, we could replace them.

**Review by santgr11 — APPROVED** [review]

File: pipelines/botbuilder-python-ci-slack-test.yml [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    SlackClientSigningSecret: $(SlackTestBotSlackClientSigningSecret)
+    SlackVerificationToken: $(SlackTestBotSlackVerificationToken)
+
+- task: AzureCLI@1
+  displayName: 'Delete resources'
+  inputs:
+    azureSubscription: $(AzureSubscription)
+    scriptLocation: inlineScript
+    inlineScript: 'call az group delete -n "$(BotGroup)" --yes'
+  condition: and(always(), ne(variables['DeleteResourceGroup'], 'false'))
```
Comment (ceciliaavila) [inline pipelines/botbuilder-python-ci-slack-test.yml:105]: We should add the `DeleteResourceGroup` variable to the list at the top to have it documented.

File: libraries/functional-tests/slacktestbot/bots/echo_bot.py [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+        if isinstance(body, SlackEvent):
+            if body.subtype == "file_share":
+                await turn_context.send_activity("Echo: I received and attachment")
+            elif body.message and body.message.attachments:
+                await turn_context.send_activity("Echo: I received a link share")
+
+    def __create_interactive_message(self, file_path: str) -> Attachment:
+        # interactive_message_json = open(file_path).read()
+        # adaptive_card_attachment = json.loads(interactive_message_json, cls=Block)
```
Comment (ceciliaavila) [inline libraries/functional-tests/slacktestbot/bots/echo_bot.py]: Can these lines be removed?

**Review by ceciliaavila — CHANGES_REQUESTED** [review]

**Comment by iscaltritti** [comment]
@ceciliaavila both changes applied, thanks for the feedback!

**Review by ceciliaavila — APPROVED** [review]

**Review by denscollo — APPROVED** [review]

**Comment by denscollo** [comment]
Promoted

**Pull request closed without merging**